flytekit/extras/tasks/shell.py

```diff
@@ -80,18 +59,81 @@ def _dummy_task_func():
 T = typing.TypeVar("T")
 
 
-class ShellTask(PythonInstanceTask[T]):
-    """ """
+class _Interpolaizer(abc.ABC):
+    @abc.abstractmethod
+    def interpolate(self, tmpl: str, inputs=None, outputs=None) -> str:
+        pass
+
+
+class _DoubleCurlyBraceInterpolizer(_Interpolaizer):
 
     _INPUT_REGEX = re.compile(r"({{\s*.inputs.(\w+)\s*}})", re.IGNORECASE)
     _OUTPUT_REGEX = re.compile(r"({{\s*.outputs.(\w+)\s*}})", re.IGNORECASE)
 
+    def interpolate(
+        self, tmpl: str, inputs: typing.Dict[str, typing.Any] = None, outputs: typing.Dict[str, typing.Any] = None
+    ) -> str:
+        inputs = inputs or {}
+        outputs = outputs or {}
+        tmpl = self._interpolate(tmpl, self._INPUT_REGEX, **inputs)
+        tmpl = self._interpolate(tmpl, self._OUTPUT_REGEX, **outputs)
+        return tmpl
+
+    @staticmethod
+    def _interpolate(tmpl: str, regex: re.Pattern, **kwargs) -> str:
+        """
+        Substitutes all templates that match the supplied regex
+        with the given inputs and returns the substituted string. The result is non destructive towards the given string.
+        """
+        modified = tmpl
+        matched = set()
+        for match in regex.finditer(tmpl):
+            expr = match.groups()[0]
+            var = match.groups()[1]
+            if var not in kwargs:
+                raise ValueError(f"Variable {var} in Query (part of {expr}) not found in inputs {kwargs.keys()}")
+            matched.add(var)
+            val = kwargs[var]
+            # str conversion should be deliberate, with right conversion for each type
+            modified = modified.replace(expr, _stringify(val))
+        return modified
+
+
+class _PythonFStringInterpolizer(_Interpolaizer):
+    class _Formatter(string.Formatter):
+        def format_field(self, value, format_spec):
+            if isinstance(value, FlyteFile):
+                value.download()
+                return value.path
+            if isinstance(value, FlyteDirectory):
+                value.download()
+                return value.path
+            if isinstance(value, datetime.datetime):
+                return value.isoformat()
+            return super().format_field(value, format_spec)
```

Comment:
Can we not call `stringify()` because the function definition looks the same?

Reply:
great question! After discussing with @kumare3 out-of-band, he suggested I get rid of the Regex based way of doing this entirely, so the stringify logic is no longer needed in two places (here and in the Regex based Interpolizer that showed up a few commits ago).  We _could_ delegate out to the stringify function within here, but since this logic is now only used in one place I think it's probably OK to just inline it here.